File: Backend/Core/src/main/scala/APIs/RegisterService.scala
package APIs

import Common.API.API
import Global.RegisterService
import Objects._

import io.circe.generic.auto._

case class CreateDepartmentMessage(
    adminToken: String,
    name: String
) extends API[Unit](RegisterService, "CreateDepartmentMessage")

case class CreateRegisterMessage(
    adminToken: String,
    studentId: String,
    departmentId: String
) extends API[Unit](RegisterService, "CreateRegisterMessage")

case class CreateTermMessage(
    adminToken: String,
    name: String
) extends API[Unit](RegisterService, "CreateTermMessage")

case class CreateTraceMessage(
    adminToken: String,
    studentId: String,
    termId: String,
    stage: Stage
) extends API[Unit](RegisterService, "CreateTraceMessage")

case class QueryDepartmentsMessage()
    extends API[List[Department]](RegisterService, "QueryDepartmentsMessage")

case class QueryTermsMessage()
    extends API[List[Term]](RegisterService, "QueryTermsMessage")

case class QueryStudentMessage(
    studentId: String
) extends API[Student](RegisterService, "QueryStudentMessage")

case class QueryStudentsMessage(
    departmentId: String
) extends API[List[User]](RegisterService, "QueryStudentsMessage")

case class UpdateDepartmentMessage(
    adminToken: String,
    departmentId: String,
    name: String
) extends API[Unit](RegisterService, "UpdateDepartmentMessage")

case class UpdateTermMessage(
    adminToken: String,
    termId: String,
    name: String
) extends API[Unit](RegisterService, "UpdateTermMessage")

case class DeleteDepartmentMessage(
    adminToken: String,
    departmentId: String
) extends API[Unit](RegisterService, "DeleteDepartmentMessage")

case class DeleteTermMessage(
    adminToken: String,
    termId: String
) extends API[Unit](RegisterService, "DeleteTermMessage")

case class DeleteRegisterMessage(
    adminToken: String,
    studentId: String,
    departmentId: String
) extends API[Unit](RegisterService, "DeleteRegisterMessage")

case class DeleteTraceMessage(
    adminToken: String,
    studentId: String,
    termId: String
) extends API[Unit](RegisterService, "DeleteTraceMessage")

File: Frontend/services/registerService.ts
import { Department, Stage, Student, Term } from "@/types/registerService";
import ApiClient from "./apiClient";

const port = 10011;
const client = new ApiClient(port);

const RegisterService = {
  createDepartment: client.api<
    {
      adminToken: string;
      name: string;
    },
    void
  >("CreateDepartmentMessage"),

  createRegister: client.api<
    {
      adminToken: string;
      studentId: string;
      departmentId: string;
    },
    void
  >("CreateRegisterMessage"),

  createTerm: client.api<
    {
      adminToken: string;
      name: string;
    },
    void
  >("CreateTermMessage"),

  createTrace: client.api<
    {
      adminToken: string;
      studentId: string;
      termId: string;
      stage: Stage;
    },
    void
  >("CreateTraceMessage"),

  queryDepartments: client.api<{ child: void }, Department[]>(
    "QueryDepartmentsMessage"
  ),

  queryTerms: client.api<{ child: void }, Term[]>("QueryTermsMessage"),

  queryStudent: client.api<{ studentId: string }, Student>(
    "QueryStudentMessage"
  ),

  queryStudents: client.api<{ departmentId: string }, Student[]>(
    "QueryStudentsMessage"
  ),

  updateDepartment: client.api<
    {
      adminToken: string;
      departmentId: string;
      name: string;
    },
    void
  >("UpdateDepartmentMessage"),

  updateTerm: client.api<
    {
      adminToken: string;
      termId: string;
      name: string;
    },
    void
  >("UpdateTermMessage"),

  deleteDepartment: client.api<
    {
      adminToken: string;
      departmentId: string;
    },
    void
  >("DeleteDepartmentMessage"),

  deleteTerm: client.api<
    {
      adminToken: string;
      termId: string;
    },
    void
  >("DeleteTermMessage"),

  deleteRegister: client.api<
    {
      adminToken: string;
      studentId: string;
      departmentId: string;
    },
    void
  >("DeleteRegisterMessage"),

  deleteTrace: client.api<
    {
      adminToken: string;
      studentId: string;
      termId: string;
    },
    void
  >("DeleteTraceMessage"),
};

export default RegisterService;
